File: backend/src/main/java/com/managestudents/auth/PortalRouting.java
package com.managestudents.auth;

import java.util.Collection;
import java.util.Locale;
import java.util.Set;
import java.util.stream.Collectors;

public final class PortalRouting {

    private PortalRouting() {
    }

    /** Chuẩn hóa mã vai trò: trim, IN HOA, bỏ tiền tố {@code ROLE_} (nếu có). */
    public static String normalizeRoleCode(String raw) {
        if (raw == null || raw.isBlank()) {
            return "";
        }
        String t = raw.trim().toUpperCase(Locale.ROOT);
        if (t.startsWith("ROLE_")) {
            return t.substring("ROLE_".length());
        }
        return t;
    }

    public static Set<String> normalizedRoleCodes(Collection<String> roleCodes) {
        if (roleCodes == null || roleCodes.isEmpty()) {
            return Set.of();
        }
        return roleCodes.stream()
                .map(PortalRouting::normalizeRoleCode)
                .filter(s -> !s.isEmpty())
                .collect(Collectors.toSet());
    }

    /** Trang mặc định sau đăng nhập: ADMIN/TEACHER → admin, còn lại → user. */
    public static String defaultRoute(Collection<String> roleCodes) {
        Set<String> up = normalizedRoleCodes(roleCodes);
        if (up.contains("ADMIN") || up.contains("TEACHER")) {
            return "/admin";
        }
        return "/user";
    }

    /**
     * Chỉ ADMIN hoặc TEACHER được vào cổng quản trị.
     */
    public static boolean canAccessAdminPortal(Collection<String> roleCodes) {
        Set<String> up = normalizedRoleCodes(roleCodes);
        if (up.isEmpty()) {
            return false;
        }
        return up.contains("ADMIN") || up.contains("TEACHER");
    }
}

File: frontend/src/lib/portalRouting.ts
/** Đồng bộ logic với backend `PortalRouting` (chuẩn hóa mã vai trò, gồm tiền tố ROLE_). */

function normalizeRoleCode(raw: string): string {
  const t = raw.trim().toUpperCase();
  if (t.startsWith("ROLE_")) return t.slice(5);
  return t;
}

export function normalizedRoleCodes(roleCodes: string[]): Set<string> {
  return new Set(
    roleCodes.map(normalizeRoleCode).filter((s) => s.length > 0)
  );
}

export function defaultRouteFromRoles(roleCodes: string[]): string {
  const up = normalizedRoleCodes(roleCodes);
  if (up.has("ADMIN") || up.has("TEACHER")) return "/admin";
  return "/user";
}

/** Chỉ ADMIN hoặc TEACHER được vào cổng quản trị. */
export function canAccessAdminPortal(roleCodes: string[]): boolean {
  const up = normalizedRoleCodes(roleCodes);
  if (up.size === 0) return false;
  return up.has("ADMIN") || up.has("TEACHER");
}
